src/test/kotlin/de/bund/digitalservice/useid/refresh/RefreshControllerTest.kt

```diff
@@ -0,0 +1,67 @@
+package de.bund.digitalservice.useid.refresh
+
+import com.ninjasquad.springmockk.MockkBean
+import de.bund.digitalservice.useid.identification.IdentificationSession
+import de.bund.digitalservice.useid.identification.IdentificationSessionService
+import io.mockk.every
+import org.junit.jupiter.api.Tag
+import org.junit.jupiter.api.extension.ExtendWith
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.Arguments
+import org.junit.jupiter.params.provider.MethodSource
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.boot.test.autoconfigure.web.reactive.WebFluxTest
+import org.springframework.security.test.context.support.WithMockUser
+import org.springframework.test.context.junit.jupiter.SpringExtension
+import org.springframework.test.web.reactive.server.WebTestClient
+import org.springframework.web.util.DefaultUriBuilderFactory
+import reactor.core.publisher.Mono
+import java.util.UUID
+import java.util.stream.Stream
+
+@Tag("test")
+@ExtendWith(value = [SpringExtension::class])
+@WebFluxTest(controllers = [RefreshController::class])
+@WithMockUser
+internal class RefreshControllerTest(@Autowired val webTestClient: WebTestClient) {
+    @MockkBean
+    private lateinit var identificationSessionService: IdentificationSessionService
+
+    private val client = clientWithoutParameterEncoding()
+
+    @ParameterizedTest
+    @MethodSource("refreshParameters")
+    fun `refresh endpoint escapes forwarded query parameters`(inputParameterString: String, expectedParameterString: String) {
+        // Given
+        val eIdSessionId: UUID = UUID.randomUUID()
+        val refreshAddress = "some-refresh-address"
+        val identificationSession = IdentificationSession(UUID.randomUUID(), refreshAddress, emptyList())
+        every { identificationSessionService.findByEIDSessionId(any()) } returns Mono.just(identificationSession)
+
+        // When
+        client
+            .get()
+            .uri("/refresh?sessionId=$eIdSessionId&$inputParameterString")
+            .exchange()
+            // Then
+            .expectStatus()
+            .is3xxRedirection
+            .expectHeader()
+            .location("$refreshAddress?sessionId=$eIdSessionId&$expectedParameterString")
+    }
+    private fun refreshParameters(): Stream<Arguments?>? {
+        return Stream.of(
+            Arguments.of("foo=alert(1)", "foo=alert%281%29"), // encode unencoded value
+            Arguments.of("alert(1)=foo", "alert%281%29=foo"), // encode unencoded key
+            Arguments.of("foo=alert%281%29", "foo=alert%281%29"), // keep value encoding
+            Arguments.of("foo=bar%26bar%3Dfoo", "foo=bar%26bar%3Dfoo"), // keep value encoding
+            Arguments.of("alert%281%29=foo", "alert%281%29=foo") // keep key encoding
```

Comment:
The result is the latter though which is right I would say. If the input is `return_to=https://www.test.de/test?foo=bar&bar=foo` it is read as two query parameters `return_to=https://www.test.de/test?foo=bar` and `bar=foo` since the URL is not encoded.

Reply:
what I was thinking, `&bar=foo` belongs to the URL not the `return_to`, so in my example there were only 1 query param which is the `return_to`, but I think it depends on the publisher if they correctly encode the URL and we are already handling it properly